android/src/test/java/com/amplitude/android/migration/RemnantDataMigrationTest.kt

```diff
@@ -0,0 +1,168 @@
+package com.amplitude.android.migration
+
+import android.content.Context
+import androidx.test.core.app.ApplicationProvider
+import com.amplitude.android.Amplitude
+import com.amplitude.android.Configuration
+import com.amplitude.core.Storage
+import kotlinx.coroutines.runBlocking
+import org.json.JSONArray
+import org.junit.Test
+import org.junit.jupiter.api.Assertions
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+import java.io.File
+import java.io.FileOutputStream
+import java.io.InputStream
+
+@RunWith(RobolectricTestRunner::class)
+class RemnantDataMigrationTest {
+    @Test
+    fun `legacy data version 4 should be migrated`() {
+        checkLegacyDataMigration("legacy_v4.sqlite", 4)
+    }
+
+    @Test
+    fun `legacy data version 3 should be migrated`() {
+        checkLegacyDataMigration("legacy_v3.sqlite", 3)
+    }
+
+    @Test
+    fun `missing legacy data should not fail`() {
+        checkLegacyDataMigration("dummy.sqlite", 0)
+    }
+
+    @Test
+    fun `no data should be migrated if migrateLegacyData=false`() {
+        checkLegacyDataMigration("legacy_v4.sqlite", 4, false)
+    }
+
+    private fun checkLegacyDataMigration(legacyDbName: String, dbVersion: Int, migrateLegacyData: Boolean = true) {
+        val context = ApplicationProvider.getApplicationContext<Context>()
+
+        val instanceName = "legacy_v${dbVersion}_$migrateLegacyData"
+        val dbPath = context.getDatabasePath("com.amplitude.api_$instanceName")
+        val inputStream = javaClass.classLoader?.getResourceAsStream(legacyDbName)
+        if (inputStream != null) {
+            copyStream(inputStream, dbPath)
+        }
+
+        val amplitude = Amplitude(
+            Configuration(
+                "test-api-key",
+                context,
+                instanceName = instanceName,
+                migrateLegacyData = migrateLegacyData,
+            )
+        )
+
+        // Check migrated data after RemnantEventsMigrationPlugin
+        runBlocking {
+            amplitude.isBuilt.await()
+
+            val identity = amplitude.idContainer.identityManager.getIdentity()
+            if (inputStream != null && migrateLegacyData) {
+                Assertions.assertEquals("22833898-c487-4536-b213-40f207abdce0R", identity.deviceId)
```

Comment:
Question: where are the original legacy data come from?  Can we use some shared test Constance instead of using real value here?

Reply:
Yes, it comes from `android/src/test/resources/legacy_v*.sqlite` files
I copied the files from my Android emulator file system.
The identifiers are unique random uids, the files do not contain api keys.